Fix coding style for mock (#806)
vdaas/vald: docs/contributing/coding-style.md

        type want struct {
            // generated test code
    ```
### Using Mock

> In Vald, we use a lot of external library, there are a lot of dependencies between libraries.

Reviewer: ```suggestion
In Vald, we use a lot of external libraries, there are a lot of dependencies between libraries.
```

As a result, due to the more complexity of the test, it has become more difficult to determine whether or not to mock dependencies.

#### Condition

When dependencies have the following factor, you can decide to mock the dependencies.

- Incomplete implementation
- I/O

>   - e.g. Network access, disk operation, or etc.

Reviewer 2: ```suggestion
  - e.g. Network access, disk operation, etc.
```

>   - e. g. Network access, disk operation, or etc.

Reviewer 2: ```suggestion
  - e. g. Network access, disk operation, etc.
```

- Hardware dependent

>   - e.g. CPU, Memory usage, disk I/O, or etc.

Reviewer 2: ```suggestion
  - e.g. CPU, Memory usage, disk I/O, etc.
```

>   - e. g. CPU, Memory usage, disk I/O, or etc.

Reviewer 2: ```suggestion
  - e. g. CPU, Memory usage, disk I/O, etc.
```

- Difficult to create error of dependencies
- Difficult to initialize

>   - e.g. Random number and time, file I/O initialization, environment dependent, or etc.

Reviewer 2: ```suggestion
  - e.g. Random number and time, file I/O initialization, environment dependent, etc.
```

>   - e. g. Random number and time, file IO initialization, environment dependent, or etc.

Reviewer 2: ```suggestion
  - e. g. Random number and time, file IO initialization, environment dependent, etc.
```

- Test result may change in each runtime

>   - e. g. Only test result may change in each runtime, System call inside implementation, or etc.

Reviewer 2: ```suggestion
  - e. g. Only test result may change in each runtime, System call inside implementation, etc.
```

#### Risk

Before applying mock to the object, you should be aware of the following risks.

> - We **do not** know whether the dependencies is correctly implemented or not.

Reviewer 2: ```suggestion
- We **do not** know whether the dependencies are correctly implemented or not.
```

> - We cannot notice the change in dependencies.

Reviewer 2: ```suggestion
- We cannot notice the changes in dependencies.
```

#### Implementation

> The implementation of mock object should be:

Reviewer 2: ```suggestion
The implementation of the mock object should be:
```

        type want struct {
            // generated test code
    ```
### Mock

In Vald, we use a lot of external library, there are a lot of dependencies between libraries.

As a result, the complexity of the test has increased, and it has become more difficult to determine whether or not to mock dependent objects.

#### Condition

> When a dependent object has the following feature, you can decide to mock the dependent.

Reviewer 2: ```suggestion
When a dependent object has the following factor, you can decide to mock the dependent object.
```

    For example, we decided to mock the following implementation `Encoder`.

    ```go
    package json

    type Encoder interface {
        Encode(interface{}) ([]byte, error) 
    }

    ```

    ```go
    type encoder struct {
        encoder json.Encoder
    }

    func (e *encoder) Encode(obj interface{}) ([]byte, error) {
        return e.encoder.Encode(obj)
    }
    ```

    The following is an example of mock implementation:

    ```go
    package json

    type MockEncoder struct {
        EncoderFunc func(interface{}) ([]byte, error)
    }

    func (m *MockEncoder) Encode(obj interface{}) ([]byte, error) {
        return m.EncodeFunc(obj)
    }

    ```

    The following is an example implementation of test code to create the mock object and mock the implementation.

    ```go
    tests := []test {
        {
            name: "returns (byte{}, nil) when encode success"
            fields: fields {
                encoding: &json.MockEncoder {
                    EncoderFunc: func(interface{}) ([]byte, error) {
                        return []byte{}, nil
                    },
                },
            }
            ......
        }
    }

    ......

    ```

> #### Risk
> 
> - We don't know if the dependent object is correct.
> - We don't notice the change in dependency.

Reviewer 2: I think it is better to move this section before ``implementation` section.
```suggestion
#### Risk

Before applying mock to the object, you should be aware of the following risks.

- We do not know if the dependent object is correctly implemented or not.
- We cannot notice the change in dependency.
```

Author: thank you.
It looks good to me.
I will fix it soon

        type want struct {
            // generated test code
    ```
### Mock

In Vald, we use a lot of external library, there are a lot of dependencies between libraries.

As a result, the complexity of the test has increased, and it has become more difficult to determine whether or not to mock dependent objects.

> #### Cases where you can use mock

Reviewer 2: ```suggestion
#### Condition
```

When a dependent object has the following feature, you can decide to mock the dependent.

- Incomplete implementation
- IO
  - Network access, disk operation, etc.
- Hardware dependent
  - CPU, memory usage, disk IO, etc.
- Difficult to create error of dependent object (when we will write error test case)
- Difficult to initialize
  - Random number and time, file IO initialization, environment dependent, etc.

> - Test result may change

Reviewer 2: ```suggestion
- Test result may change in runtime
```

  - Cases where the implementation and test code are not changed but the test result changes

>   - System call dependent

Reviewer 2: ```suggestion
  - System call inside implementation
```

> #### How to create mock

Reviewer 2: ```suggestion
#### Implementation
```

1. Basic mock example.

    For example, we decided to mock the following implementation `Encoder`.

    ```go
    package json

>     type Encoder struct {

Reviewer 2: is it an interface?

Author: thank you. I made a misstake. so I will fix it soon.

        Encode(interface{}) ([]byte, error) 
    }

    ```

    ```go
    type encoder struct {
        encoder json.Encoder
    }

    func (e *encoder) Encode(obj interface{}) ([]byte, error) {
        return e.encoder.Encode(obj)
    }
    ```

    And there are 4 rules of mock code.
    

>     - File location is same pacakge as mock target.

Reviewer 2: ```suggestion
    - Same package as the mock target.
```

    - File name is `〇〇_mock.go`

>     - Structure name are `Mock{Interface name}`

Reviewer 2: ```suggestion
    - Struct name is `Mock{Interface name}`
```

>     - Method injected from test code is `{Method name}Func`

Reviewer 2: We should move these rules to another section

>     - Method injected from test code is `{Method name}Func`

Reviewer 2: ```suggestion
```

Author: I think we should write about mock method format.
what do you think?

Reviewer 2: I think so.
But it is not the method name, it is the function implementation of the functions inside the mock object

    The following is an example of mock implementation:

    ```go
    package json

    type MockEncoder struct {
        EncoderFunc func(interface{}) ([]byte, error)
    }

    func (m *MockEncoder) Encode(obj interface{}) ([]byte, error) {
        return m.EncodeFunc(obj)
    }

    ```

    The following is an example implementation of test code to create the mock object and mock the implementation.

    ```go

> 

Reviewer 2: ```suggestion
```

                },
            }
            ......
        }
    }

    ......

    ```

#### Rules of mock code

> There are 4 rules of mock code.

Reviewer 2: ```suggestion
There are 3 rules of mock code.
```

> #### Rules of mock code
> 
> There are 4 rules of mock code.

Reviewer 2: Please move this category inside `implementation` section
```suggestion
The implementation of mock object should be:
```

Author: thank you. 

by the way, It is different from the following opinion, so I would like you to unify it from now on.
https://github.com/vdaas/vald/pull/806#discussion_r513967570

- Same package as the mock target.
- File name is `〇〇_mock.go`

> - Struct name is `Mock{Interface name}`

Reviewer 2: I think we should describe the rule first before the example

Author: I think so too. thankyou

> #### Mock rick

Reviewer 2: I think this section should before the example
```suggestion
#### Risk
```

> - We don't know if it's really correct.

Reviewer 2: can you explain more? I cannot understand what it means

> 1. Basic mock example.
> 
>     For example, we decided to mock the following implementation `Encoder`.
> 
>     ```go
>     package json
> 
>     type Encoder interface {
>         Encode(interface{}) ([]byte, error) 
>     }
> 
>     ```
> 
>     ```go
>     type encoder struct {
>         encoder json.Encoder
>     }
> 
>     func (e *encoder) Encode(obj interface{}) ([]byte, error) {
>         return e.encoder.Encode(obj)
>     }
>     ```
> 
>     The following is an example of mock implementation:
> 
>     ```go
>     package json
> 
>     type MockEncoder struct {
>         EncoderFunc func(interface{}) ([]byte, error)
>     }
> 
>     func (m *MockEncoder) Encode(obj interface{}) ([]byte, error) {
>         return m.EncodeFunc(obj)
>     }
> 
>     ```
> 
>     The following is an example implementation of test code to create the mock object and mock the implementation.
> 
>     ```go
>     tests := []test {
>         {
>             name: "returns (byte{}, nil) when encode success"
>             fields: fields {
>                 encoding: &json.MockEncoder {
>                     EncoderFunc: func(interface{}) ([]byte, error) {
>                         return []byte{}, nil
>                     },
>                 },
>             }
>             ......
>         }
>     }
> 
>     ......
> 
>     ```

Reviewer: ```suggestion
  For example, we decided to mock the following implementation `Encoder`.

  ```go
  package json

  type Encoder interface {
      Encode(interface{}) ([]byte, error) 
  }

  ```

  ```go
  type encoder struct {
      encoder json.Encoder
  }

  func (e *encoder) Encode(obj interface{}) ([]byte, error) {
      return e.encoder.Encode(obj)
  }
  ```

  The following is an example of mock implementation:

  ```go
  package json

  type MockEncoder struct {
      EncoderFunc func(interface{}) ([]byte, error)
  }

  func (m *MockEncoder) Encode(obj interface{}) ([]byte, error) {
      return m.EncodeFunc(obj)
  }

  ```

  The following is an example implementation of test code to create the mock object and mock the implementation.

  ```go
  tests := []test {
      {
          name: "returns (byte{}, nil) when encode success"
          fields: fields {
              encoding: &json.MockEncoder {
                  EncoderFunc: func(interface{}) ([]byte, error) {
                      return []byte{}, nil
                  },
              },
          },
          ......
      }
  }

  ```

```
